client/src/action/auth.js

```diff
@@ -0,0 +1,83 @@
+
+import setAuthToken from "../utils/setAuthToken";
+import axios from "axios";
+import {createAsyncThunk} from "@reduxjs/toolkit";
+import {showAlert} from "./alert";
+import {authError, loginSuccess, registerFail, registerSuccess, userLoaded} from "../reducers/auth";
+
+//load user
+export const loadUser = () => async (dispatch) => {
+    if (localStorage.token) {
+        setAuthToken(localStorage.token);
+    }
+    try {
+        const res = await axios.get("/api/auth");
+        dispatch(userLoaded());
+    } catch (err) {
+        dispatch(authError());
+    }
+};
+
+//Register User
+export const registerUser = createAsyncThunk(
+    " auth/registerUser",
+    async (formData, {dispatch, rejectWithValue}) => {
+        try {
+            const res = await axios.post("/api/users", formData);
+            console.log(res, "res");
+            dispatch(registerSuccess(res.data));
+
+            dispatch(loadUser());
+
+            localStorage.setItem("token", res.data.token);
+            return res.data;
+        } catch (err) {
+            dispatch(registerFail());
+            if (err.response && err.response.data && err.response.data.errors) {
+                const errors = err.response.data.errors;
+                errors.forEach((error) => {
+                    if (error.param === "email") {
+                        dispatch(showAlert({type: "danger", message: "Email is invalid"}));
+                    } else if (error.param === "password") {
+                        dispatch(
+                            showAlert({type: "danger", message: "Password is invalid"})
+                        );
+                    } else {
+                        dispatch(showAlert({type: "danger", message: error.msg}));
+                    }
+                });
+            } else {
+                dispatch(
+                    showAlert({type: "danger", message: "An error occurred"})
+                );
+            }
+            return rejectWithValue(err.response?.data);
+        }
+    }
+);
+
+//login user
+export const loginUser = createAsyncThunk(
+    " auth/loginUser",
+    async (formData, {dispatch, rejectWithValue}) => {
+        try {
+            const res = await axios.post("/api/auth", formData);
+            dispatch(loginSuccess(res.data));
```

Comment:
Here is the where the problem was. You were previously dispatching actions incorrectly (using redux tool kit)
Here is what you were doing before. We just need to dispatch `loginSuccess(...)`
```
dispatch(dispatch({type: "loginSuccess" , payload: res.data});
```
We do not need action types using RTK, so we can just dispatct the name of the action